components/role-mgt/org.wso2.carbon.identity.role.v2.mgt.core/src/main/java/org/wso2/carbon/identity/role/v2/mgt/core/RoleManagementServiceImpl.java

```diff
@@ -0,0 +1,666 @@
+    private static final Log audit = CarbonConstants.AUDIT_LOG;
+    private final RoleDAO roleDAO = RoleMgtDAOFactory.getInstance().getRoleDAO();
+    private static final String auditMessage
+            = "Initiator : %s | Action : %s | Target : %s | Data : { %s } | Result : %s ";
+    private final String success = "Success";
+
+    @Override
+    public RoleBasicInfo addRole(String roleName, List<String> userList, List<String> groupList,
+                                 List<Permission> permissions, String audience, String audienceId, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        if (StringUtils.startsWithIgnoreCase(roleName, UserCoreConstants.INTERNAL_SYSTEM_ROLE_PREFIX)) {
+            String errorMessage = String.format("Invalid role name: %s. Role names with the prefix: %s, is not allowed"
+                            + " to be created from externally in the system.", roleName,
+                    UserCoreConstants.INTERNAL_SYSTEM_ROLE_PREFIX);
+            throw new IdentityRoleManagementClientException(INVALID_REQUEST.getCode(), errorMessage);
+        }
+        if (isDomainSeparatorPresent(roleName)) {
+            // SCIM2 API only adds roles to the internal domain.
+            throw new IdentityRoleManagementClientException(INVALID_REQUEST.getCode(), "Invalid character: "
+                    + UserCoreConstants.DOMAIN_SEPARATOR + " contains in the role name: " + roleName + ".");
+        }
+
+        // Validate audience.
+        if (StringUtils.isNotEmpty(audience)) {
+            if (!(ORGANIZATION.equalsIgnoreCase(audience) || APPLICATION.equalsIgnoreCase(audience))) {
+                throw new IdentityRoleManagementClientException(INVALID_AUDIENCE.getCode(), "Invalid role audience");
+            }
+            if (ORGANIZATION.equalsIgnoreCase(audience)) {
+                validateOrganizationRoleAudience(audienceId);
+                audience = ORGANIZATION;
+            }
+            if (APPLICATION.equalsIgnoreCase(audience)) {
+                validateApplicationRoleAudience(audienceId, tenantDomain);
+                audience = APPLICATION;
+            }
+        } else {
+            audience = ORGANIZATION;
+            audienceId = getOrganizationIdByTenantDomain(tenantDomain);
+        }
+
+        RoleManagementEventPublisherProxy roleManagementEventPublisherProxy = RoleManagementEventPublisherProxy
+                .getInstance();
+        roleManagementEventPublisherProxy.publishPreAddRoleWithException(roleName, userList, groupList, permissions,
+                audience, audienceId, tenantDomain);
+        RoleBasicInfo roleBasicInfo = roleDAO.addRole(roleName, userList, groupList, permissions, audience, audienceId,
+                tenantDomain);
+        roleManagementEventPublisherProxy.publishPostAddRole(roleBasicInfo.getId(), roleName, userList, groupList,
+                permissions, audience, audienceId, tenantDomain);
```

Comment:
On successful role addition with the application audience, need to invalidate the application cache.
Need to do this via event handler

Reply:
same apply for the role deletion and role name update

Reply:
is this done?